djtriggers/models.py

```diff
@@ -75,35 +82,111 @@ def __init__(self, *args, **kwargs):
 
+    def save(self, *args, **kwargs):
+        # Save the content type of this object, so we can retrieve it later when querying Trigger
+        if not self.content_type:
+            self.content_type = ContentType.objects.get_for_model(self.__class__)
+        super(Trigger, self).save(*args, **kwargs)
+
+    def as_leaf_class(self):
+        """
+        Return this object as an instance of the Trigger subclass it was created as
+        """
+        model = self.content_type.model_class()
+        if model == Trigger:
+            return self
+        return model.objects.get(id=self.id)
+
     def set_source(self, *args):
         args = [str(arg) for arg in args]
         self.source = '$'.join(args)
 
     def get_source(self):
         return tuple(x for x in self.source.split('$') if x != '')
 
-    def process(self, force=False, logger=None, dictionary={}):
-        if logger:
-            self.logger = get_logger(logger)
-        now = timezone.now()
-        if not force and self.date_processed is not None:
-            raise AlreadyProcessedError()
-        if not force and self.process_after and self.process_after >= now:
-            raise ProcessLaterError(self.process_after)
-
+    def process(self, force=False, logger=None, dictionary=None, use_statsd=False):
+        """
+        Executes the Trigger
+        :param bool force: force the execution
+        :param string logger: slug of preferred logger
+        :param dict dictionary: dictionary needed by trigger to execute
+        :param bool use_statsd: whether to use statsd
+        :return: None
+        """
+        if dictionary is None:
+            dictionary = {}
+
+        # The task gets locked because multiple tasks in the queue can process the same trigger.
+        # The lock assures no two tasks can process a trigger simultaneously.
+        # The check for date_processed assures a trigger is not executed multiple times.
         try:
-            self.logger.log_result(self, self._process(dictionary))
-        except ProcessLaterError as e:
-            self.process_after = e.process_after
-            self.save()
-            raise
-        if self.date_processed is None:
-            self.date_processed = now
-        self.save()
+            with NonBlockingLock.objects.acquire_lock(self):
+                if logger:
+                    self.logger = get_logger(logger)
+                now = timezone.now()
+                if not force and self.date_processed is not None:
+                    raise AlreadyProcessedError()
+                if not force and self.process_after and self.process_after >= now:
+                    raise ProcessLaterError(self.process_after)
+
+                try:
+                    # execute trigger
+                    self.logger.log_result(self, self._process(dictionary))
+                    self._handle_execution_success(use_statsd)
+                except ProcessLaterError as e:
```

Comment:
The `_process` function of the trigger itself may raise it, if it decided that the `process_after` date was incorrect and it wants to be executed later.